float64 output_x
float64 output_y
float64 output_z
float64 output_w